web/controllers/expertController/index.ts

```diff
@@ -0,0 +1,328 @@
+import {Request, Response} from "express";
+import {Student} from "../../../common/entity/Student";
+import {getLogger} from "log4js";
+import {Pupil} from "../../../common/entity/Pupil";
+import {ApiContactExpert, ApiGetExpert, ApiGetExpertiseTag, ApiPutExpert} from "./format";
+import {getTransactionLog} from "../../../common/transactionlog";
+import {getManager} from "typeorm";
+import {ExpertData} from "../../../common/entity/ExpertData";
+import mailjet from "../../../common/mails/mailjet";
+import {DEFAULTSENDERS} from "../../../common/mails/config";
+import ContactExpertEvent from "../../../common/transactionlog/types/ContactExpertEvent";
+import {ExpertiseTag} from "../../../common/entity/ExpertiseTag";
+
+const logger = getLogger();
+
+/**
+ * @api {POST} /expert/:id/contact contactExpert
+ * @apiVersion 1.1.0
+ * @apiDescription
+ * Writes an email to an expert
+ *
+ * If email was successfully sent, status code 200 is returned.
+ * Note: delivery cannot be guaranteed.
+ *
+ * @apiName ContactExpert
+ * @apiGroup Expert
+ *
+ * @apiUse Authentication
+ * @apiUse ContentType
+ *
+ * @apiUse ContactExpert
+ *
+ * @apiParam (URL Parameter) {string} id Expert Id
+ *
+ * @apiExample {curl} Curl
+ * curl -k -i -X POST -H "Token: <AUTHTOKEN>" -H "Content-Type: application/json" https://[HOST]/api/expert/:id/contact
+ *
+ * @apiUse StatusOk
+ * @apiUse StatusBadRequest
+ * @apiUse StatusNotFound
+ * @apiUse StatusInternalServerError
+ */
+export async function postContactExpertHandler(req: Request, res: Response) {
+    let status = 200;
+    try {
+        if ((res.locals.user instanceof Student || res.locals.user instanceof Pupil)
+            && req.params.id != undefined
+            && typeof req.body.emailText == 'string') {
+
+            if (req.body.emailText.length === 0) {
+                logger.warn(`Empty email text specified when trying to send mentor-contact-mail!`);
+                status = 400;
+            }
+
+            if (status < 300) {
+                status = await postContactExpert(req.params.id, res.locals.user, req.body);
+            }
+        } else {
+            status = 400;
+            logger.warn("Invalid request for POST /expert/:id/contact");
+            logger.debug(req.body);
+        }
+    } catch (e) {
+        logger.error("POST /expert/:id/contact failed with ", e.message);
+        logger.debug(req, e);
+        status = 500;
+    }
+    res.status(status).end();
+}
+
+async function postContactExpert(id: string, user: Pupil | Student, apiContactExpert: ApiContactExpert): Promise<number> {
+    const entityManager = getManager();
+    const transactionLog = getTransactionLog();
+
+    const expert = await entityManager.findOne(ExpertData, { id: Number(id) });
```

Comment:
We cannot be sure that `id` is a string that really contains a number. If not, this will result in `NaN` and a failure.